akka-stream/src/main/scala/akka/stream/impl/FoldResourceSinkAsync.scala

```diff
@@ -0,0 +1,162 @@
+
+    val createLogic = new GraphStageLogic(shape) with InHandler {
+      lazy val decider = inheritedAttributes.mandatoryAttribute[SupervisionStrategy].decider
+
+      private implicit def ec = materializer.executionContext
+
+      private var blockingStream: Option[S] = None
+
+      private var writeFail: Option[Throwable] = None
+
+      private val createdCallback = getAsyncCallback[Try[S]] {
+        case Success(resource) ⇒
+          blockingStream = Some(resource)
+          pull(in)
+
+        case Failure(t) ⇒ failStage(t)
+      }.invokeWithFeedback _
+
+      private val errorHandler: PartialFunction[Throwable, Unit] = {
+        case NonFatal(ex) ⇒ decider(ex) match {
+          case Supervision.Stop ⇒
+            writeFail = Some(ex)
+            failStage(ex)
+          case Supervision.Restart ⇒ restartResource()
+          case Supervision.Resume  ⇒ pull(in)
+        }
+      }
+
+      private val pushCallback = getAsyncCallback[Try[Unit]] {
+        case Success(_) ⇒ pull(in)
+        case Failure(t) ⇒ errorHandler(t)
+      }.invoke _
+
+      override def preStart(): Unit = createResource()
+
+      override def onPush(): Unit =
+        blockingStream match {
+          case Some(resource) ⇒
+            try {
+              write(resource, grab(in)).onComplete(pushCallback)(sameThreadExecutionContext)
+            } catch errorHandler
+          case None ⇒
+          // we got a pull but there is no open resource, we are either
+          // currently creating/restarting then the pull will be triggered when creating the
+          // resource completes, or shutting down and then the push does not matter anyway
+        }
+
+      override def postStop(): Unit = {
+        val o = blockingStream
+        blockingStream = None
+        o match {
+          case Some(s) ⇒
+            writeFail.foreach(failStage)
+
+            try {
+              val f = close(s)
+              writeFail match {
+                case Some(ex) ⇒ promise.tryCompleteWith(f.flatMap { _ ⇒ Future.failed(ex) })
+                case None ⇒ promise.tryCompleteWith(f.map(_ ⇒ Done).recoverWith {
+                  case NonFatal(ex) ⇒
+                    failStage(ex)
+                    Future.failed(ex)
+                })
+              }
+            } catch {
+              case NonFatal(ex) ⇒
+                if (writeFail.isEmpty) failStage(ex)
+                promise.tryFailure(writeFail.getOrElse(ex))
+            }
+
+          case None ⇒ writeFail match {
+            case None     ⇒ promise.trySuccess(Done)
+            case Some(ex) ⇒ promise.tryFailure(ex)
+          }
+        }
+      }
+
+      private def restartResource(): Unit = {
+        val o = blockingStream
+        blockingStream = None
+
+        o match {
+          case Some(resource) ⇒
+            // wait for the resource to close before restarting
+            try {
+              close(resource).onComplete(getAsyncCallback[Try[Unit]] {
```

Comment:
If the stage stops while we are closing, and closing fails that is currently lost.

Reply:
Okay I'll have to think about this one